core/trino-main/src/main/java/io/trino/sql/planner/optimizations/BeginTableWrite.java

```diff
@@ -225,6 +235,29 @@ private WriterTarget createWriterTarget(WriterTarget target)
             throw new IllegalArgumentException("Unhandled target type: " + target.getClass().getSimpleName());
         }
 
+        private TableHandle findTableScanHandle(PlanNode node)
```

Comment:
The logic here seems duplicated over `rewriteModifyTableScan`.
Can we somehow avoid the duplication?

it looks tongue-in-cheek to traverse the plan to find the TH from the TableScan so that we can call beginDelete and traverse the plan back and put a new TH into the TableScan. 

Perhaps instead of adding new method, remove the TH param from `rewriteModifyTableScan` and instead pass some information or function, what to do when TableScan is eventually found.

Reply:
I agree with your observation, but these methods are called from different places. `getTarget` is called from `visitTableFinish` and `rewriteModifyTableScan` is called from `visitDelete`. With the current plan structure, the target is stored in the `TableFinishNode`, but we modify the table scan in the `DeleteNode`, using the target saved in the visitor context.

There's probably a way to clean this up, but that's outside the scope of fixing this bug.